File: python/Problem13.py
# Problem: 13. Roman to Integer
# URL: https://leetcode.com/problems/roman-to-integer/
# Difficulty: Easy
# Tags: Map

from typing import Dict

class Solution:
  def romanToInt(self, s: str) -> int:
    length: int = len(s)

    romanToNums: Dict = {
      'I': 1,
      'V': 5,
      'X': 10,
      'L': 50,
      'C': 100,
      'D': 500,
      'M': 1000
    }

    result: int = 0
    for i in range(0, length):
      if i + 1 < length and romanToNums[s[i]] < romanToNums[s[i + 1]]:
        result -= romanToNums[s[i]]
      else:
        result += romanToNums[s[i]]
    return result

File: cpp/Problem13.cpp
/*
 * Problem: 13. Roman to Integer
 * URL: https://leetcode.com/problems/roman-to-integer/
 * Difficulty: Easy
 * Tags: Map
 */

#include <string>

using namespace std;

class Solution {
public:
	int romanToInt(string s) {
		int len = s.length();
		
		int* values = new int[len];
		char romanNumerals[7] = {'I', 'V', 'X', 'L', 'C', 'D', 'M'};
		int romanToNumbers[7] = {1, 5, 10, 50, 100, 500, 1000};

		for (int i = 0; i < len; i++) {
			for (int j = 0; j < 7; j++) {
				if (s[i] == romanNumerals[j]) {
					values[i] = romanToNumbers[j];
					break;
				}
			}
		}

		int result = 0;
		for (int i = 0; i < len; i++) {
			if (i + 1 < len && values[i] < values[i + 1]) {
				result -= values[i];
			} else {
				result += values[i];
			}
		}
		delete[] values;
		return result;
	}
};